### PR TITLE
Laboratory work #4, Merkulova Ekaterina - 19FPL2

### PR DESCRIPTION
**Pull request by katerinamerkulova**

### CODE REVIEW

File: .pylintrc [hunk 1]
```diff
@@ -142,7 +142,9 @@ disable=print-statement,
         comprehension-escape,
         bad-continuation,
         unused-argument,
-        unnecessary-pass
+        unnecessary-pass,
+        too-few-public-methods,
+        protected-access
```
Comment (dmitry-uraev) [inline .pylintrc:147]: good solution, acceptable

File: lab_4/backoff_text_generator_score_ten_test.py [hunk 2]
```diff
@@ -157,7 +157,7 @@ def test_generate_next_word_incorrect_context(self):
         two = NGramTrie(2, encoded)
         four = NGramTrie(4, encoded)
 
-        bad_inputs = [[], {}, (2000, 1000, ), None, 9, 9.34, True]
+        bad_inputs = ([], {}, tuple(), None, 9, 9.34, True)
```
Comment (dmitry-uraev) [inline lab_4/backoff_text_generator_score_ten_test.py]: yes, as we have discussed it

File: lab_4/decode_text_score_eight_test.py [hunk 3]
```diff
@@ -62,7 +62,7 @@ def test_decode_text_incorrect_storage(self):
 
         to_decode = generator.generate_text(context, 2)
 
-        bad_inputs = [(), [], 123, None, NGramTrie]
+        bad_inputs = (tuple(), [], {}, 123, None, NGramTrie)
```
Comment (dmitry-uraev) [inline lab_4/decode_text_score_eight_test.py]: why explicit tuple() notation? why not just ()?

File: lab_4/likelihood_text_generator_score_eight_test.py [hunk 4]
```diff
@@ -157,7 +157,7 @@ def test_generate_next_word_incorrect_context(self):
 
         trie = NGramTrie(3, encoded)
 
-        bad_inputs = [[], {}, (2000, 1000, ), None, 9, 9.34, True]
+        bad_inputs = ([], {}, tuple(), (2000, ), None, 9, 9.34, True)
```
Comment (dmitry-uraev) [inline lab_4/likelihood_text_generator_score_eight_test.py]: and why not list, but tuple?

File: lab_4/main.py [hunk 5]
```diff
@@ -2,73 +2,306 @@
...
+from lab_4 import validation
 
 
 def tokenize_by_sentence(text: str) -> tuple:
-    pass
+    validation.ensure_type((text, str))
+    validation.ensure_not_empty((text, ))
+
+    clean_text = re.sub(r'[^.!?\w\s]', '', text)
+    tokens = tuple(re.sub(r'([A-Z][\w\s]+)[.!?]?',
```
Comment (dmitry-uraev) [inline lab_4/main.py:22]: good

File: lab_4/main.py [hunk 6]
```diff
@@ -2,73 +2,306 @@
...
+import re
+
+from lab_4.ngrams.ngram_trie import NGramTrie
+from lab_4 import validation
 
 
 def tokenize_by_sentence(text: str) -> tuple:
-    pass
+    validation.ensure_type((text, str))
+    validation.ensure_not_empty((text, ))
```
Comment (dmitry-uraev) [inline lab_4/main.py:19]: interesting, but i like decorators more)

File: lab_4/main.py [hunk 7]
```diff
@@ -2,73 +2,306 @@
...
+                'n_grams': trie.n_grams,
+                'size': trie.size,
+                'uni_grams': trie.uni_grams,
+                }
+            _n_gram_tries.append(_n_gram_trie)
+
+        model_dict['data']['_n_gram_tries'] = _n_gram_tries
+
+    with open(path_to_saved_model, 'w') as file:
+        pprint(model_dict, stream=file, compact=True)
```
Comment (dmitry-uraev) [inline lab_4/main.py:263]: nice

File: lab_4/main.py [hunk 8]
```diff
@@ -2,73 +2,306 @@
...
+                trie['encoded_text'],
+                )
+            _n_gram_trie.n_gram_frequencies = trie['n_gram_frequencies']
+            _n_gram_trie.n_grams = trie['n_grams']
+            _n_gram_trie.uni_grams = trie['uni_grams']
+            tries.append(_n_gram_trie)
+
+        generator = BackOffGenerator(_word_storage, _n_gram_trie, *tries)
+
+    return generator
```
Comment (dmitry-uraev) [inline lab_4/main.py:307]: let's discuss it on Saturday, I like it

**Review by dmitry-uraev — COMMENTED** [review]
Great job! Done here

File: lab_4/decode_text_score_eight_test.py (hunk 3 above)
Comment (katerinamerkulova) [reply lab_4/decode_text_score_eight_test.py]: oh, you're right, i'll fix it))

File: lab_4/likelihood_text_generator_score_eight_test.py (hunk 4 above)
Comment (katerinamerkulova) [reply lab_4/likelihood_text_generator_score_eight_test.py]: for consistency of a whole project, including main.py

File: lab_4/main.py (hunk 6 above)
Comment (katerinamerkulova) [reply lab_4/main.py:19]: me too, but it became too custom for each function, so i decided to do it more flexible

**Pull request closed without merging**